### PR TITLE
[SPARK-8390][Streaming][Kafka] fix docs related to HasOffsetRanges

### PR DESCRIPTION
**Pull request by koeninger**

### CODE REVIEW

**Comment by koeninger** [comment]
I don't think a doc change caused a python test failure.

File: docs/streaming-kafka-integration.md [hunk 1]
```diff
@@ -74,15 +74,15 @@ Next, we discuss how to use this approach in your streaming application.
...
 
-- *Efficiency:* Achieving zero-data loss in the first approach required the data to be stored in a Write Ahead Log, which further replicated the data. This is actually inefficient as the data effectively gets replicated twice - once by Kafka, and a second time by the Write Ahead Log. This second approach eliminate the problem as there is no receiver, and hence no need for Write Ahead Logs.
+- *Efficiency:* Achieving zero-data loss in the first approach required the data to be stored in a Write Ahead Log, which further replicated the data. This is actually inefficient as the data effectively gets replicated twice - once by Kafka, and a second time by the Write Ahead Log. This second approach eliminates the problem as there is no receiver, and hence no need for Write Ahead Logs.  As long as you have sufficient Kafka retention, messages can be recovered from Kafka.
```
Comment (tdas) [inline docs/streaming-kafka-integration.md]: Are there two spaces before "As long as"

File: docs/streaming-kafka-integration.md [hunk 2]
```diff
@@ -74,15 +74,15 @@ Next, we discuss how to use this approach in your streaming application.
...
+- *Exactly-once semantics:* The first approach uses Kafka's high level API to store consumed offsets in Zookeeper. This is traditionally the way to consume data from Kafka. While this approach (in combination with write ahead logs) can ensure zero data loss (i.e. at-least once semantics), there is a small chance some records may get consumed twice under some failures. This occurs because of inconsistencies between data reliably received by Spark Streaming and offsets tracked by Zookeeper. Hence, in this second approach, we use simple Kafka API that does not use Zookeeper.  Offsets are tracked by Spark Streaming within its checkpoints. This eliminates inconsistencies between Spark Streaming and Zookeeper/Kafka, and so each record is received by Spark Streaming effectively exactly once despite failures.  In order to achieve exactly-once semantics for output of your results, your save method must be either idempotent, or an atomic transaction that saves results and offsets in your own data store.
```
Comment (tdas) [inline docs/streaming-kafka-integration.md]: Again are there double spaces after periods or its I am just seeing it wrong.

**Comment by tdas** [comment]
If the java example is updated to the access the offsets, then the scala example should also be updated.

**Comment by koeninger** [comment]
Yeah, the spacing in that document in general is a mess (mix of tabs and spaces, some 2 spaces between sentences, etc).  I cleaned it up somewhat.  Also further fixed the scala / java offset ranges examples, the java one is pretty much a C&P from the test at this point.

File: docs/streaming-kafka-integration.md [hunk 3]
```diff
@@ -74,15 +74,15 @@ Next, we discuss how to use this approach in your streaming application.
...
+- *Exactly-once semantics:* The first approach uses Kafka's high level API to store consumed offsets in Zookeeper. This is traditionally the way to consume data from Kafka. While this approach (in combination with write ahead logs) can ensure zero data loss (i.e. at-least once semantics), there is a small chance some records may get consumed twice under some failures. This occurs because of inconsistencies between data reliably received by Spark Streaming and offsets tracked by Zookeeper. Hence, in this second approach, we use simple Kafka API that does not use Zookeeper. Offsets are tracked by Spark Streaming within its checkpoints. This eliminates inconsistencies between Spark Streaming and Zookeeper/Kafka, and so each record is received by Spark Streaming effectively exactly once despite failures. In order to achieve exactly-once semantics for output of your results, your save method must be either idempotent, or an atomic transaction that saves results and offsets in your own data store.
```
Comment (tdas) [inline docs/streaming-kafka-integration.md]: * "save method" --> "output operation that saves the data to external data stores" 
* Also refer to the fault-tolerance section in the main programming guide. This stuff is common for everything.
*nit: Can you capitalize Write Ahead Logs

File: docs/streaming-kafka-integration.md [hunk 4]
```diff
@@ -161,6 +161,8 @@ Next, we discuss how to use this approach in your streaming application.
 
 	You can use this to update Zookeeper yourself if you want Zookeeper-based Kafka monitoring tools to show progress of the streaming application.
 
-	Another thing to note is that since this approach does not use Receivers, the standard receiver-related (that is, [configurations](configuration.html) of the form `spark.streaming.receiver.*` ) will not apply to the input DStreams created by this approach (will apply to other input DStreams though). Instead, use the [configurations](configuration.html) `spark.streaming.kafka.*`. An important one is `spark.streaming.kafka.maxRatePerPartition` which is the maximum rate at which each Kafka partition will be read by this direct API. 
+	Note that the typecast to HasOffsetRanges will only succeed if it is done in the first method called on the directKafkaStream, not later down a chain of methods. You can use transform() instead of foreachRDD() as your first method call in order to access offsets, then call further Spark methods. However, be aware that the one-to-one mapping between RDD partition and Kafka partition does not remain after any methods that shuffle or repartition, e.g. reduceByKey() or window().
```
Comment (tdas) [inline docs/streaming-kafka-integration.md]: Good addition. :)

**Comment by tdas** [comment]
Could you update the example to something that accesses the offsets in the foreachRDD? 
```
var offsetRanges: OffsetRange[] = _
...
kafkaStream.transform { rdd => 
   offsetRanges = rdd.asInstanceOf[HasOffsetRanges].offsetRanges
   rdd
}.map { ... }.reduceByKey { ... }.foreachRDD { rdd => 
   // use offsetRanges to do something, maybe print the ranges
}
```

Do the same for Java and Scala

**Comment by tdas** [comment]
So the JIRA was about updating the examples actually. Its great that you have updated the docs AND the tests, but it would ideal if the examples DirectKafkaWordCount and JavaDirectKafkaWordCount are updated to show how the offset ranges can be accessed. Since you have updated the tests, mind updating the examples as well?

**Comment by koeninger** [comment]
The word count examples don't have any need of accessing offsets.

Wouldn't it be better to have separate examples?  I don't want someone
thinking they need to do all this typecast hoop jumping just to get a word
count

**Comment by tdas** [comment]
That is a good point. Then make two different examples one for Java and
Scala.
BTW, I would like to have this for 1.4.1, so gotta merge soon. So I think I
will merge this, and you can make another PR with the same JIRA to add the
new examples.

**Comment by tdas** [comment]
Could you update title to get the ordering right? [JIRA][Streaming][Kafka]. Otherwise LGTM.

**Comment by koeninger** [comment]
Title should be right now

**Comment by tdas** [comment]
I have merged this to master and branch-1.4

**Pull request closed without merging**